service/src/main/java/bio/terra/cbas/dependencies/sam/SamService.java

```diff
@@ -19,6 +27,35 @@ private StatusApi getStatusApi() {
     return new StatusApi(samClient.getApiClient());
   }
 
+  public UsersApi getUsersApi(String accessToken) {
+    return new UsersApi(samClient.getApiClient(accessToken));
+  }
+
+  public String getUserToken() {
+    // RequestContextHolder exposes the web request in the form of a *thread-bound*
+    // [RequestAttributes](https://docs.spring.io/spring-framework/docs/current/javadoc-api/org/springframework/web/context/request/RequestAttributes.html) object.
+    return (String)
+        RequestContextHolder.currentRequestAttributes()
+            .getAttribute(BearerTokenFilter.ATTRIBUTE_NAME_TOKEN, RequestAttributes.SCOPE_REQUEST);
+  }
+
+  // this cache uses token.hashCode as its key. This prevents any logging such as
+  // in CacheLogger from logging the raw token.
+  // Taken from WDS
+  @Cacheable(cacheNames = "tokenResolution", key = "#root.target.getUserToken().hashCode()")
```

Comment:
Not sure how concerned we should be about `hashCode` collisions... Wondering if there is a better way we can encode the token to avoid logging raw token but without the potential for collisions.

Reply:
I think I'd feel better if the cached function took arguments like `(String token, String tokenHash)` and it was _that_ hash that was the key to the cache. It should be easy enough to wrap the call in a way that lets you extract the relevant fields and it would reduce the amount of magic (like... what is "`root`"?)

By the way, in java `hashCode`, unless overridden, gives you a value depending on memory address, so this might not always give the same hashcode for the same "value".

Reply:
Looks like `BearerToken.hashCode()` uses `Objects.hash`, which should be stable for Strings